File: Problems/python/boj_과자나눠주기.py
#https://www.acmicpc.net/problem/16401
#Binary Search

import sys
input = sys.stdin.readline
import bisect

m, n = map(int, input().split())
l = list(map(int, input().split()))
l.sort()

s, e = 1, l[-1]
result = 0
while s<=e:
    mid = (s+e)//2
    idx = bisect.bisect_left(l, mid)
    
    cnt = 0
    for i in range(idx, n): #과자는 자를 수 있으니 이렇게 카운트 해야한다.
        cnt += l[i]//mid
        
    if cnt>=m:
        s = mid+1
        result = mid
    else:
        e = mid-1

print(result)

File: Problems/cpp/boj_과자나눠주기.cpp
//https://www.acmicpc.net/problem/16401
//Binary Search

#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

int m, n;
vector<int> snack;

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    cin >> m >> n;

    snack.resize(n);
    for (int i=0; i<n; i++)
        cin >> snack[i];

    int s = 1;
    int e = *max_element(snack.begin(), snack.end());

    int result = 0;
    while (s<=e){
        int count = 0;
        int mid = (s+e)/2;

        for (const int s : snack)
            if (mid<=s) count += s/mid;

        if (count>=m){
            s = mid + 1;
            result = mid;
        }
        else
            e = mid - 1;
    }

    cout << result;
}